flumeDevice

```diff
@@ -222,7 +223,8 @@ def fetchNotifications()
 
     if(resp?.data)
     {
-        def tempCount = getNotifCount() ?: 0
+        def tempCount = getNotifCount() ?: 0        
+        def processingStale = (tempCount == 0)
```

Comment:
logic on handling non-stale makes sense. What I am having trouble figuring out is how `processingStale = (tempCount == 0)` defines staleness.

NotificationCount is set as `setNotifCount(tempCount)` which is the number of notifications processed. How is it ever `0` again? I might just be overlooking a piece of the code that Im new to.

Reply:
tempCount gets set to 0 when the State Variables don't have a value at all.  This represents the state after a fresh Configure(), which happens at initial install and when the settings are Updated().  If there's a non-zero number returned by getNotifCount(), then there seems to definitely be a notification that we want to be sure to process.

Reply:
but how do you know it wasnt already processed, thats the part tripping me up I think

Reply:
oh, or was the case that needed to be solved only `when the settings are Updated()` ... which then, this all makes sense. I thought historic notifications were being processed continually, and that was where I didnt understand how you knew if a notification in the `resp` was already processed

Reply:
Ah, got it.  I track the number of processed notifications in state with the driver.

When they are queried from Flume, all from the history get returned, and the duplicate ones need to be dropped.  So, if there's a non-zero number of past notifications, then anything beyond that is actually interesting on the Hubitat side because it hasn't been processed yet.

Unless you Updated() and got set back to 0 in state, in which case we don't know whether they're stale or not.  So, I just drop them all.

Check my logic on all of that.

Reply:
ok, this makes sense and your logic seems solid 
```
Unless you Updated() and got set back to 0 in state, in which case we don't know whether they're stale or not. So, I just drop them all.
```

Now .. I need to ask about the latter 

```
    if(resp?.data)
    {
        def tempCount = getNotifCount() ?: 0

        if(resp.data.size() >= tempCount)
        {
            // this is a typical state;
            //   there are at least as many notifications as we think we already have read

            // hack the known ones off of the beginning, then process the rest
            for(i = 0; i < tempCount; i++)
            {
                resp.data.remove(0)
            }
        }
        else
        {
            // less are in the notification than we think we already have read,
            //  so they must have been cleared in the app

            // reset count and process them all to catch up
            tempCount = 0
        }

        for(notification in resp.data)
        {
            tempCount++

            sendEvent(name: "notificationStream", value: "[${notification.created_datetime}]: ${notification.message}", isStateChange: true)

            // short pause to try to keep them sequential in the event log
            pauseExecution(10)
        }
        setNotifCount(tempCount)
    }
```

Is `tempCount` only ever the new number of `resp` via `tempCount++` ... which you slice down via `resp.data.remove(0)` ... maybe Im just misreading this? As its been working ??

For example, say initial grab of notificatins was `20`, you process them. tempCount is `20`
You grab it again and its `21`, you slice out the original `20`, tempCount is `1` 

Now you grab `22` and you only slice out `1` ? This is where i might be lost

I sort of feel logic should be

```
def priorCount = getNotifCount() ?: 0

slice out priorCount amount
process new resp
setNotifCount(tempCount + priorCount) (or simpler setNotifCount(size(resp))

```

on a `update()`

```
processStale = true (current logic)
setNotifCount(size(resp)). <--- to reset the known processed notifications?
```

Reply:
oh, NM .. I see whats going on. The tempCount = 0 is when there are less, and you have no idea when/if notifications were cleared. So you just start over. 

Does this not introduce duplicates as well?

I think everything is solid, not that I understand the logic